### DISCUSSION TITLE
Mutable

### DISCUSSION THREAD

**Discussion by lacasaprivata2** [tokio-rs/tokio#4205] [Q&A]
https://github.com/tokio-rs/tokio/blob/d1a400912e82505c18c6c0c1f05cda06f334e201/tokio/src/sync/mpsc/unbounded.rs#L126

Why does this need to be mutable when it's known to be accessed from 1 thread / send() is not mutable?

**Comment by Darksonn** [comment]
The mpsc channel is a multi-producer single-consumer channel, and having the `recv` methods take `&self` would allow several threads to receive in parallel, in contradiction with it being a single-consumer channel.

The `send` method can take `&self` because having multiple threads send in parallel is fine.

**Reply by lacasaprivata2** [reply]
Hm - thought this would be up to the programmer to verify

how would you recommend going about storing receivers in a global, static vector for a single worker thread to poll?

**Reply by Darksonn** [reply]
If you wrap the receiver (or the vector with receivers) in a mutex, you can do it. Otherwise you can also use an mpmc channel, which is available in e.g. the async-channel crate.
